--- ./wpa_supplicant/config.h.old	2013-01-12 16:42:53.000000000 +0100
+++ ./wpa_supplicant/config.h	2014-10-22 10:58:17.394959248 +0200
@@ -797,6 +797,17 @@
 	 * this default behavior.
 	 */
 	enum mfp_options pmf;
+
+#ifdef CONFIG_CMDSTATE
+        /**
+	 * change_state_cmd - Exec to call when wpa_supplicant change state
+	 *
+	 *   This file is called when wpa_supplicant change state
+	 *   wpa_supplicant fork and execve with
+	 *   informations in the environnement.
+	 */
+        char *change_state_cmd;
+#endif
 };
 
 
