--- wpa_supplicant/config.h.old	2013-08-26 16:32:30.424137951 +0200
+++ wpa_supplicant/config.h	2013-08-26 16:58:40.074514621 +0200
@@ -797,6 +797,15 @@
 	 * this default behavior.
 	 */
 	enum mfp_options pmf;
+
+        /**
+	 * event_cmd - Path to the event file
+	 *
+	 *   This file is called after a wpa_supplicant event
+	 *   wpa_supplicant fork and execve event file with
+	 *   information in the environnement.
+	 */
+        char *event_cmd;
 };
 
 
