--- ./wpa_supplicant/wpa_supplicant_i.h.old	2013-01-12 16:42:53.000000000 +0100
+++ ./wpa_supplicant/wpa_supplicant_i.h	2014-10-22 11:48:36.931209851 +0200
@@ -41,6 +41,12 @@
 struct ctrl_iface_global_priv;
 struct wpas_dbus_priv;
 
+#include <sys/wait.h>
+
+#ifdef CONFIG_CMDSTATE
+#include "cmdstate.h"
+#endif
+
 /**
  * struct wpa_interface - Parameters for wpa_supplicant_add_iface()
  */
@@ -246,6 +252,9 @@
 #define MAX_WFD_SUBELEMS 10
 	struct wpabuf *wfd_subelem[MAX_WFD_SUBELEMS];
 #endif /* CONFIG_WIFI_DISPLAY */
+#ifdef CONFIG_CMDSTATE
+        struct wpa_cmd cmdstate_exe;
+#endif
 };
 
 
