--- ./wpa_supplicant/wpa_supplicant.c.old	2013-01-12 16:42:53.000000000 +0100
+++ ./wpa_supplicant/wpa_supplicant.c	2014-10-22 12:00:19.663424489 +0200
@@ -707,6 +707,10 @@
 		    old_state == WPA_COMPLETED)
 			wpas_notify_auth_changed(wpa_s);
 	}
+#ifdef CONFIG_CMDSTATE
+	if (wpa_s->conf->change_state_cmd != NULL)
+	  wpa_supplicant_callcmd(wpa_s, state);
+#endif
 }
 
 
@@ -3223,6 +3227,11 @@
 	global = os_zalloc(sizeof(*global));
 	if (global == NULL)
 		return NULL;
+
+#ifdef CONFIG_CMDSTATE
+	wpa_supplicant_init_exec(global);
+#endif
+
 	dl_list_init(&global->p2p_srv_bonjour);
 	dl_list_init(&global->p2p_srv_upnp);
 	global->params.daemonize = params->daemonize;
@@ -3378,6 +3387,9 @@
 	os_free(global->params.override_driver);
 	os_free(global->params.override_ctrl_interface);
 
+#ifdef CONFIG_CMDSTATE
+	wpa_supplicant_deinit_exec(&(global->cmdstate_exe));
+#endif
 	os_free(global->p2p_disallow_freq);
 
 	os_free(global);
