--- wpa_supplicant/config.h.old	2013-07-05 11:51:15.631244976 +0200
+++ wpa_supplicant/config.h	2013-07-05 11:55:08.098975253 +0200
@@ -345,6 +345,16 @@
 	 *   1 = only include configured SSIDs in scan results/BSS table
 	 */
 	int filter_ssids;
+
+
+	/**
+	 * event_cmd - Path to the event file
+	 *
+	 *   This file is called after a wpa_supplicant event
+	 *   wpa_supplicant fork and execve event file with
+	 *   information in the environnement.
+	 */
+	char *event_cmd;
 };
 
 
